sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,284 @@
+            labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+            # Update medoids with the new cluster indices
+            self._update_medoid_idxs_in_place(distances, labels, medoid_idxs)
+
+        # Expose labels_ which are the assignments of
+        # the training data to clusters
+        self.labels_ = labels
+        self.cluster_centers_ = X[medoid_idxs]
+        self.inertia_ = self._compute_inertia(X)
+
+        # Return self to enable method chaining
+        return self
+
+    def _update_medoid_idxs_in_place(self, distances, cluster_idxs,
+                                     medoid_idxs):
+        """In-place update of the medoid indices"""
+
+        # Update the medoids for each cluster
+        for k in range(self.n_clusters):
+
+            if sum(cluster_idxs == k) == 0:
+                warnings.warn("Cluster %d is empty!" % k)
+                continue
+
+            # Extract the distance matrix between the data points
+            # inside the cluster k
+            cluster_k_idxs = np.where(cluster_idxs == k)[0]
+            in_cluster_distances = distances[np.ix_(cluster_k_idxs,
+                                                    cluster_k_idxs)]
+
+            # Calculate all costs from each point to all others in the cluster
+            in_cluster_all_costs = np.sum(in_cluster_distances, axis=1)
+
+            min_cost_idx = np.argmin(in_cluster_all_costs)
+            min_cost = in_cluster_all_costs[min_cost_idx]
+            curr_cost = np.sum(distances[medoid_idxs[k], cluster_k_idxs])
+
+            # If the minimum cost is smaller than that
+            # exhibited by the currently used medoid,
+            # we switch to using the new medoid in cluster k
+            if min_cost < curr_cost:
+                # Find data points that belong to cluster k,
+                # and assign the newly found medoid as the medoid
+                # for cluster k
+                medoid_idxs[k] = cluster_k_idxs[min_cost_idx]
+
+    def transform(self, X):
+        """Transforms X to cluster-distance space.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=(n_samples, n_features)
+            Data to transform.
+
+        Returns
+        -------
+        X_new : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            X transformed in the new space of distances to cluster centers.
+        """
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        check_is_fitted(self, "cluster_centers_")
+
+        return pairwise_distances(X, Y=self.cluster_centers_,
+                                  metric=self.distance_metric)
+
+    def predict(self, X):
+        """Predict the closest cluster for each sample in X
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            New data to predict.
+
+        Returns
+        -------
+        labels : array, shape [n_samples,]
```

Comment:
please go over all you new doctrings to match recommended format

http://scikit-learn.org/stable/developers/contributing.html#random-numbers